day3/recipe/RecipesModel.h

```diff
@@ -0,0 +1,59 @@
+#ifndef RECIPESMODEL_H
+#define RECIPESMODEL_H
+
+#include <QtCore>
+#include <QtGui>
+#include <memory>
+
+
+class RecipesModel : public QAbstractListModel
+{
+    Q_OBJECT
+public:
+    enum DifficultyLevel {
+        LowDifficulty,
+        MidDifficulty,
+        HighDifficulty,
+        DifficultyCount,
+    };
+    Q_ENUM(DifficultyLevel)
+
+    static const QString difficultyStrings[];
+
+    enum RoleNames {
+        DescriptionRole = Qt::UserRole,
+        ImageSourceRole = Qt::UserRole + 1,
+        DifficultyRole = Qt::UserRole + 2,
+        PreparationTimeRole = Qt::UserRole + 3,
+        DifficultyStringRole = Qt::UserRole + 4,
+        PreferredRole = Qt::UserRole + 5
+    };
+
+    struct RecipeData {
+        RecipeData(QString descr, QString  imageSo, DifficultyLevel difficul, int prepTime, bool pref): description(descr), imageSource(imageSo), difficulty(difficul)
+            , preparationTime(prepTime), preferred(pref){}
+
+        QString  description;
+        QString  imageSource;
+        DifficultyLevel difficulty;
+        int preparationTime;
+        bool preferred;
+    };
+
+    explicit RecipesModel(QObject *parent = 0);
+    ~RecipesModel();
+
+public: // QAbstractItemModel interface
+    int rowCount(const QModelIndex &parent) const override;
+    QVariant data(const QModelIndex &index, int role) const override;
+    bool setData(const QModelIndex& index, const QVariant& value, int role = Qt::EditRole) override;
+    Qt::ItemFlags flags(const QModelIndex& index) const override;
+protected:
+    QHash<int, QByteArray> roleNames() const override;
+private:
+    QList<std::shared_ptr<RecipeData>> m_data;
```

Comment:
QList internamente memorizza gli oggetti come `T` se sono di dimensioni pari o inferiore alla dimensione di un puntatore, altrimenti memorizza i puntatori agli indirizzi degli oggetti:
```
Internally, QList<T> is represented as an array of T if sizeof(T) <= sizeof(void*) and T has been declared to be either a Q_MOVABLE_TYPE or a Q_PRIMITIVE_TYPE using Q_DECLARE_TYPEINFO. Otherwise, QList<T> is represented as an array of T* and the items are allocated on the heap.
```

Nel caso di `std::shared_ptr` la dimensione è pari a 16Byte (dovrebbe essere un contatore atomico e un puntatore), contro gli 8Byte di un `void*` (architettura 64bit).
Questo significa che QList ti sta memorizzando puntatori a `std::shared_ptr` anzichè `std::shared_ptr`, quindi c'è un ulteriore livello di indirezione!

Se usi QList, ricordati che ci pensa lui a ottimizzare per le copie e spostamenti, non devi farlo tu.

Reply:
rimossi i pointers, adesso usiamo un vector che ha come template direttamente le struct _"RecipeData"_ `QVector<RecipeData> m_data`